doc/tests.rst

```diff
@@ -448,8 +449,151 @@ which displays nicer looking diffs that the standard ``diff``
 tool. You can change that by passing ``--diff-command CMD`` to
 dune.
 
+Cram Tests
+==========
+
+Cram tests are expectation tests written in a shell like syntax. They are ideal
+for testing binaries. Cram tests are auto discovered from files or directories
+with a ``.t`` extension, so they must be enabled manuallly in the
+``dune-project`` file:
+
+.. code:: scheme
+
+   (lang dune 2.7)
+   (cram enable)
+
+
+File Tests
+----------
+
+To define a standalone test, we create a ``.t`` file. For example, ``foo.t``:
+
+.. code:: bash
+   Simplest possible cram test
+     $ echo "testing"
+
+This simple example demonstrates two components of cram tests:
+
+* Comments - Anything that doesn't start with a 2 space indentation is a comment
+
+* Commands - A command starts with 2 spaces followed by a ``$``. It is executed
+  in the shell and the output is diffed against the output below. In this
+  example, there's no output yet.
+
+To run the test and promote the results:
+
+.. code:: bash
+   $ dune runtest
+   $ dune promote
+
+We now see the output of the command:
+
+.. code:: bash
+   Simplest possible cram test
+     $ echo "testing"
+     testing
+
+This is the main advantage of expect tests. We don't need to write assertions
+manually, instead we detect failure when the command produces a different output
+than what is recorded in the test script.
+
+For example, here's an example of how we'd test the ``wc`` utility. ``wc.t``:
+.. code:: bash
+
+   We create a fixture called foo
+     $ cat >foo <<EOF
+     > foo
+     > bar
+     > baz
+     > EOF
+
+   After creating the fixture, we want to verify that ``wc`` gives us the right
+   result:
+     $ wc -l foo | awk '{ print $1 }'
+     4
+
+The above example uses the here doc syntax to pipe the subsequent lines to
+``cat``. This is convenient for creating small fixtures.
+
+Directory Tests
+---------------
+
+In the above example we used ``cat`` to create the fixture, but what if there
+are too man fixtures to comfortably fit in test file? Or some of the fixtures
+are binary? It's possible to include the fixtures as normal files or directories
+provided the test is defined as a directory. The name of the test directory must
+end with ``.t`` and must include a ``run.t`` as the test script. Everything else
+in that directory is treated as raw data for the test. It's not possible to
+define rules using ``dune`` files in such a directory.
+
+We convert the ``wc`` test above into a directory test ``wc.t``:
+
+.. code:: bash
+
+   $ ls wc.t
+     run.t foo.txt bar/
+
+This defines a directory test ``wc.t`` with ``run.t`` as the test script, with
```

Comment:
Reading this, it's not clear if `run.t` is a special name or if the test can be named differently.